waiter/src/waiter/scheduler/composite.clj

```diff
@@ -0,0 +1,128 @@
+  (:require [clojure.string :as str]
+            [plumbing.core :as pc]
+            [schema.core :as s]
+            [waiter.scheduler :as scheduler]
+            [waiter.util.utils :as utils]))
+
+(defrecord CompositeScheduler [service-id->scheduler scheduler-id->scheduler]
+
+  scheduler/ServiceScheduler
+
+  (get-service->instances [_]
+    (->> (vals scheduler-id->scheduler)
+         (map #(future (scheduler/get-service->instances %)))
+         (map deref)
+         (apply merge)))
+
+  (get-services [_]
+    (->> (vals scheduler-id->scheduler)
+         (map #(future (scheduler/get-services %)))
+         (map deref)
+         (apply concat)))
+
+  (get-instances [_ service-id]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/get-instances service-id)))
+
+  (kill-instance [_ instance]
+    (-> instance
+        :service-id
+        service-id->scheduler
+        (scheduler/kill-instance instance)))
+
+  (service-exists? [_ service-id]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/service-exists? service-id)))
+
+  (create-service-if-new [_ descriptor]
+    (-> descriptor
+        :service-id
+        service-id->scheduler
+        (scheduler/create-service-if-new descriptor)))
+
+  (delete-service [_ service-id]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/delete-service service-id)))
+
+  (scale-service [_ service-id target-instances force]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/scale-service service-id target-instances force)))
+
+  (retrieve-directory-content [_ service-id instance-id host directory]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/retrieve-directory-content service-id instance-id host directory)))
+
+  (service-id->state [_ service-id]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/service-id->state service-id)))
+
+  (state [_]
+    (pc/map-vals scheduler/state scheduler-id->scheduler)))
+
+(defn service-id->scheduler
+  "Resolves the scheduler for a given service-id using the scheduler defined in the description.
+   If the scheduler is not specified, the default scheduler is returned."
+  [service-id->service-description-fn scheduler-id->scheduler default-scheduler-id service-id]
+  (let [service-description (service-id->service-description-fn service-id)
+        scheduler-id (get service-description "scheduler" default-scheduler-id)]
+    (if-let [scheduler (scheduler-id->scheduler scheduler-id)]
+      scheduler
+      (throw (ex-info "Invalid scheduler specified!"
+                      {:available-schedulers (-> scheduler-id->scheduler keys sort)
+                       :service-id service-id
+                       :specified-scheduler scheduler-id})))))
+
+(defn invoke-component-factory
+  "Creates a component based on the factory-fn specified in the component-config."
+  [context {:keys [factory-fn] :as component-config}]
+  (if factory-fn
```

Comment:
Doesn't the `component-schema` validation make this condition superfluous?

Reply:
Yes, will remove.